backends/ebpf/psa/externs/ebpfPsaTableImplementation.cpp

```diff
@@ -206,4 +207,444 @@ void EBPFActionProfilePSA::applyImplementation(CodeBuilder* builder, cstring tab
+            if (left != right)
+                printError = true;
+        }
+    } else {
+        printError = true;
+    }
+
+    if (printError) {
+        ::error(ErrorType::ERR_EXPECTED,
+                "%1%: selector type keys list differs from previous %2% "
+                "(tables use the same implementation %3%)",
+                instance->table->container, table->container, declaration);
+    }
+}
+
+void EBPFActionSelectorPSA::verifyTableEmptyGroupAction(const EBPFTablePSA * instance) {
+    auto iega = instance->table->container->properties->getProperty("psa_empty_group_action");
+
+    if (emptyGroupAction == nullptr && iega == nullptr)
+        return;  // nothing to do here
+    if (emptyGroupAction == nullptr && iega != nullptr) {
+        ::error(ErrorType::ERR_UNEXPECTED,
+                "%1%: property not specified in previous table %2% "
+                "(tables use the same implementation %3%)",
+                iega, table->container, declaration);
+        return;
+    }
+    if (emptyGroupAction != nullptr && iega == nullptr) {
+        ::error(ErrorType::ERR_EXPECTED,
+                "%1%: missing property %2%, defined in previous table %3% "
+                "(tables use the same implementation %4%)", instance->table->container,
+                emptyGroupAction, table->container->toString(), declaration);
+        return;
+    }
+
+    bool same = true;
+    cstring additionalNote;
+
+    if (emptyGroupAction->isConstant != iega->isConstant) {
+        same = false;
+        additionalNote = "; note: const qualifiers also must be the same";
+    }
+
+    // compare action and arguments
+    auto rev = iega->value->to<IR::ExpressionValue>()->expression;
+    auto lev = emptyGroupAction->value->to<IR::ExpressionValue>()->expression;
+    auto rpe = rev->to<IR::PathExpression>();
+    auto lpe = lev->to<IR::PathExpression>();
+    auto rmce = rev->to<IR::MethodCallExpression>();
+    auto lmce = lev->to<IR::MethodCallExpression>();
+
+    if (lpe != nullptr && rpe != nullptr) {
+        if (lpe->toString() != rpe->toString())
+            same = false;
+    } else if (lmce != nullptr && rmce != nullptr) {
+        if (lmce->method->to<IR::PathExpression>()->path->name.originalName !=
+            rmce->method->to<IR::PathExpression>()->path->name.originalName) {
+            same = false;
+        } else if (lmce->arguments->size() == rmce->arguments->size()) {
+            for (size_t i = 0; i < lmce->arguments->size(); ++i) {
+                if (lmce->arguments->at(i)->expression->toString() !=
+                    rmce->arguments->at(i)->expression->toString()) {
+                    same = false;
+                    additionalNote = "; note: action arguments must be the same for both tables";
+                }
+            }
+        } else {
+            same = false;
+        }
+    } else {
+        same = false;
+        additionalNote = "; note: action name can\'t be mixed with "
```

Comment:
I thought that the midend or frontend enforces this.

Reply:
Unfortunately, not. This check is refers to situation where two tables specifiy `psa_empty_group_action` property with shared ActionSelector. First table can be specified as `psa_empty_group_action = NoAction;` (translated into `IR::PathExpression`) and a second table `psa_empty_group_action = NoAction();` (translated into `IR::MethodCallExpression`). The P4 code is valid for both tables and the compiler will reach this place.

I've not found a pass which could convert custom table property into `MethodCallExpression`. This is done for `default_action` but in frontend here: https://github.com/p4lang/p4c/blob/1794c3b46dbbe332e0c3727186c04959553e73c0/frontends/p4/createBuiltins.cpp#L38-L48

To do this it might be useful to write new mindend pass to rewrite such properties, but I propose to postpone this to the future PR.

Reply:
Yes, a new midend pass could be a good idea.